File: lib/api/articles.ts
import { apiClient } from "../api-client"

export interface Article {
  id: string
  title: string
  slug: string
  category: string
  content: string
  excerpt: string
  author: string
  read_time: string
  published_at: string
  is_published: boolean
  created_at: string
  updated_at: string
}

export interface ArticleSummary {
  id: string
  title: string
  slug: string
  category: string
  excerpt: string
  author: string
  read_time: string
  published_at: string
}

export interface ArticleListResponse {
  total: number
  items: ArticleSummary[]
}

export interface ArticleCreate {
  title: string
  category: string
  content: string
  excerpt: string
  author: string
  read_time: string
  is_published?: boolean
}

export interface ArticleUpdate {
  title?: string
  category?: string
  content?: string
  excerpt?: string
  author?: string
  read_time?: string
  is_published?: boolean
}

export const articlesApi = {
  /**
   * Get all articles with pagination
   */
  getAll: async (params?: {
    skip?: number
    limit?: number
    category?: string
    is_published?: boolean
  }): Promise<ArticleListResponse> => {
    const searchParams = new URLSearchParams()
    if (params?.skip !== undefined) searchParams.append("skip", params.skip.toString())
    if (params?.limit !== undefined) searchParams.append("limit", params.limit.toString())
    if (params?.category) searchParams.append("category", params.category)
    if (params?.is_published !== undefined) searchParams.append("is_published", params.is_published.toString())

    return apiClient<ArticleListResponse>(`/articles?${searchParams.toString()}`)
  },

  /**
   * Get article by slug
   */
  getBySlug: async (slug: string): Promise<Article> => {
    return apiClient<Article>(`/articles/slug/${slug}`)
  },

  /**
   * Get articles by category
   */
  getByCategory: async (category: string, skip = 0, limit = 20): Promise<ArticleListResponse> => {
    return apiClient<ArticleListResponse>(`/articles/category/${category}?skip=${skip}&limit=${limit}`)
  },

  /**
   * Get related articles
   */
  getRelated: async (articleId: string, limit = 3): Promise<ArticleSummary[]> => {
    return apiClient<ArticleSummary[]>(`/articles/${articleId}/related?limit=${limit}`)
  },

  /**
   * Create article (admin only)
   */
  create: async (data: ArticleCreate): Promise<Article> => {
    return apiClient<Article>("/articles", {
      method: "POST",
      body: JSON.stringify(data),
    })
  },

  /**
   * Update article (admin only)
   */
  update: async (id: string, data: ArticleUpdate): Promise<Article> => {
    return apiClient<Article>(`/articles/${id}`, {
      method: "PATCH",
      body: JSON.stringify(data),
    })
  },

  /**
   * Delete article (admin only)
   */
  delete: async (id: string): Promise<void> => {
    return apiClient(`/articles/${id}`, {
      method: "DELETE",
    })
  },
}

File: ruqya-backend/app/api/v1/articles.py
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.orm import Session
from typing import Optional
import uuid
from slugify import slugify

from app.database import get_db
from app.models.article import Article
from app.models.user import User
from app.schemas.article import (
    ArticleCreate,
    ArticleUpdate,
    ArticleResponse,
    ArticleSummaryResponse,
    ArticleListResponse,
    ArticleRelatedResponse,
)
from app.core.security import get_current_admin_user

router = APIRouter(prefix="/articles", tags=["Articles"])


@router.get("", response_model=ArticleListResponse)
async def get_articles(
    skip: int = Query(0, ge=0),
    limit: int = Query(20, ge=1, le=100),
    category: Optional[str] = None,
    is_published: Optional[bool] = Query(True),
    db: Session = Depends(get_db)
):
    """Get all articles with optional filtering."""
    query = db.query(Article)
    
    if is_published is not None:
        query = query.filter(Article.is_published == is_published)
    
    if category:
        query = query.filter(Article.category == category)
    
    total = query.count()
    articles = query.order_by(Article.published_at.desc()).offset(skip).limit(limit).all()
    
    return ArticleListResponse(
        total=total,
        items=[ArticleSummaryResponse.model_validate(a) for a in articles]
    )


@router.get("/slug/{slug}", response_model=ArticleResponse)
async def get_article_by_slug(slug: str, db: Session = Depends(get_db)):
    """Get an article by slug."""
    article = db.query(Article).filter(Article.slug == slug).first()
    
    if not article:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Article not found"
        )
    
    if not article.is_published:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Article not found"
        )
    
    return ArticleResponse.model_validate(article)


@router.get("/category/{category}", response_model=ArticleListResponse)
async def get_articles_by_category(
    category: str,
    skip: int = Query(0, ge=0),
    limit: int = Query(20, ge=1, le=100),
    db: Session = Depends(get_db)
):
    """Get articles by category."""
    query = db.query(Article).filter(
        Article.category == category,
        Article.is_published == True
    )
    
    total = query.count()
    articles = query.order_by(Article.published_at.desc()).offset(skip).limit(limit).all()
    
    return ArticleListResponse(
        total=total,
        items=[ArticleSummaryResponse.model_validate(a) for a in articles]
    )


@router.post("", response_model=ArticleResponse, status_code=status.HTTP_201_CREATED)
async def create_article(
    article_data: ArticleCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user)
):
    """Create a new article (Admin only)."""
    # Generate slug from title
    base_slug = slugify(article_data.title)
    slug = base_slug
    counter = 1
    
    # Ensure slug is unique
    while db.query(Article).filter(Article.slug == slug).first():
        slug = f"{base_slug}-{counter}"
        counter += 1
    
    new_article = Article(
        id=str(uuid.uuid4()),
        slug=slug,
        **article_data.model_dump()
    )
    
    db.add(new_article)
    db.commit()
    db.refresh(new_article)
    
    return ArticleResponse.model_validate(new_article)


@router.patch("/{article_id}", response_model=ArticleResponse)
async def update_article(
    article_id: str,
    article_data: ArticleUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user)
):
    """Update an article (Admin only)."""
    article = db.query(Article).filter(Article.id == article_id).first()
    
    if not article:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Article not found"
        )
    
    # Update fields
    update_data = article_data.model_dump(exclude_unset=True)
    
    # If title is being updated, regenerate slug
    if "title" in update_data:
        base_slug = slugify(update_data["title"])
        slug = base_slug
        counter = 1
        
        while db.query(Article).filter(
            Article.slug == slug,
            Article.id != article_id
        ).first():
            slug = f"{base_slug}-{counter}"
            counter += 1
        
        article.slug = slug
    
    for field, value in update_data.items():
        setattr(article, field, value)
    
    db.commit()
    db.refresh(article)
    
    return ArticleResponse.model_validate(article)


@router.delete("/{article_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_article(
    article_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user)
):
    """Delete an article (Admin only)."""
    article = db.query(Article).filter(Article.id == article_id).first()
    
    if not article:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Article not found"
        )
    
    db.delete(article)
    db.commit()
    
    return None


@router.get("/{article_id}/related", response_model=list[ArticleRelatedResponse])
async def get_related_articles(
    article_id: str,
    limit: int = Query(3, ge=1, le=10),
    db: Session = Depends(get_db)
):
    """Get related articles based on category."""
    article = db.query(Article).filter(Article.id == article_id).first()
    
    if not article:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Article not found"
        )
    
    related = db.query(Article).filter(
        Article.category == article.category,
        Article.id != article_id,
        Article.is_published == True
    ).limit(limit).all()
    
    return [ArticleRelatedResponse.model_validate(a) for a in related]
    